File: src/main/java/org/zwobble/mammoth/docx/ZippedDocxFile.java
package org.zwobble.mammoth.docx;

import java.io.IOException;
import java.io.InputStream;
import java.util.Optional;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

public class ZippedDocxFile implements DocxFile {
    private final ZipFile zipFile;

    public ZippedDocxFile(ZipFile zipFile) {
        this.zipFile = zipFile;
    }

    @Override
    public Optional<InputStream> tryGetInputStream(String name) throws IOException {
        ZipEntry entry = zipFile.getEntry(name);
        if (entry == null) {
            return Optional.empty();
        } else {
            return Optional.of(zipFile.getInputStream(entry));
        }
    }

    @Override
    public void close() throws IOException {
        zipFile.close();
    }
}

File: dotnet/Mammoth/Couscous/mammoth/internal/docx/ZippedDocxFile.cs
using System.IO.Compression;
using Mammoth.Couscous.java.io;
using Mammoth.Couscous.java.util;

namespace Mammoth.Couscous.org.zwobble.mammoth.@internal.docx {
    internal class ZippedDocxFile : DocxFile {
        private readonly ZipArchive _zipFile;
        
        internal ZippedDocxFile(File file) {
            _zipFile = ZipFile.OpenRead(file.Path);
        }

        internal ZippedDocxFile(System.IO.Stream stream) {
            _zipFile = new ZipArchive(stream);
        }
        
        public Optional<InputStream> tryGetInputStream(string name) {
            var entry = _zipFile.GetEntry(name);
            if (entry == null) {
                return Optional.empty<InputStream>();
            } else {
                return Optional.of(ToJava.StreamToInputStream(entry.Open()));
            }
        }
        
        public void close() {
            _zipFile.Dispose();
        }
    }
}
